kimchi/src/circuits/polynomials/foreign_field_add/circuitgates.rs

```diff
@@ -0,0 +1,233 @@
+//~
+//~ The upper-bound check can be calculated as
+//~ - `bound_lo = result_lo - foreign_modulus_lo - bound_carry_lo * 2^{88}`
+//~ - `bound_mi = result_mi - foreign_modulus_mi - bound_carry_mi * 2^{88} + bound_carry_lo`
+//~ - `bound_hi = result_hi - foreign_modulus_hi + 2^{88} + bound_carry_mi`
+//~
+//~ `bound_carry_i` $= 0$ or $1$ or $-1$ are auxiliary variables that handle carries between limbs
+//~
+//~ The range check of `bound` can be skipped until the end of the operations
+//~ and `result` is an intermediate value that is unused elsewhere (since the final `result`
+//~ must have had the right number of moduli subtracted along the way).
+//~
+//~ You could lay this out as a double-width gate for chained foreign additions and a final row, e.g.
+//~
+//~ | col | `ForeignFieldAdd`       | more `ForeignFieldAdd` | or `ForeignFieldFin` |
+//~ | --- | ----------------------- | ---------------------- | -------------------- |
+//~ |   0 | `left_input_lo`  (copy) | `result_lo` (copy)     | `resmin_lo` (copy)   |
+//~ |   1 | `left_input_mi`  (copy) | `result_mi` (copy)     | `resmin_mi` (copy)   |
+//~ |   2 | `left_input_hi`  (copy) | `result_hi` (copy)     | `resmin_hi` (copy)   |
+//~ |   3 | `right_input_lo` (copy) |  ...                   | `bound_lo`  (copy)   |
+//~ |   4 | `right_input_mi` (copy) |  ...                   | `bound_mi`  (copy)   |
+//~ |   5 | `right_input_hi` (copy) |  ...                   | `bound_hi`  (copy)   |
+//~ |   6 | `field_overflow`        |  ...                   |  -                   |
+//~ |   7 | `carry_lo`              |  ...                   | `bound_carry_lo`     |
+//~ |   8 | `carry_mi`              |  ...                   | `bound_carry_mi`     |
+//~ |   9 | `sign`                  |  ...                   |  -                   |
+//~ |  10 |                         |                        |                      |
+//~ |  11 |                         |                        |                      |
+//~ |  12 |                         |                        |                      |
+//~ |  13 |                         |                        |                      |
+//~ |  14 |                         |                        |                      |
+//~
+//~ Having a specific final row that checks the bound is useful as it checks the upper bound
+//~ without reusing the same constraints in the `ForeignFieldAdd` rows (which would require
+//~ some extra constraints to be added to the circuit).
+
+/// Implementation of the ForeignFieldAddition gate
+pub struct ForeignFieldAdd<F>(PhantomData<F>);
+
+impl<F> Argument<F> for ForeignFieldAdd<F>
+where
+    F: FftField,
+{
+    const ARGUMENT_TYPE: ArgumentType = ArgumentType::Gate(GateType::ForeignFieldAdd);
+    const CONSTRAINTS: u32 = 7;
+
+    fn constraint_checks<T: ExprOps<F>>(env: &ArgumentEnv<F, T>) -> Vec<T> {
+        let field_overflow = env.witness_curr(6);
+        let sign = env.witness_curr(9);
+
```

Comment:
Maybe put above
```rust
// Field overflow and sign constraints
```

Reply:
I mean not only, because later you have the ones of the carry and the result

Reply:
but I can certainly write that comment on top of the vector